### ISSUE TITLE
Support screenshots comparison

### ISSUE THREAD

**Issue by ildarsharafutdinov** [mmcc007/screenshots#65]
Do you plan to support screenshot comparison?
That would allow to write screenshot tests.

**Comment by mmcc007** [comment]
Do you mean something similar to this feature for widget tests https://api.flutter.dev/flutter/flutter_test/matchesGoldenFile.html but for integration tests?

Can you explain more about how you want it to work?

**Comment by ildarsharafutdinov** [comment]
@mmcc007 ,

Yes, something like goldens.

I think, screenshot images make sense when they're recorded in a context which is close to the production one. E.g. we need a way to use real fonts(goldens screenshots look like this): 

![](https://raw.githubusercontent.com/ksheremet/golden_test/382abee2662982f0b2b569ba5979103aea8f24fd/test/main.png)


`screenshots` currently runs driver tests on emulator and records screenshots, however it cannot compare them to reference images. 

What I mean is something like:

1. run tests in a `record` mode to save reference pngs and fail the tests
2. run tests in a `comparison` mode to make screenshots and compare to the reference pngs

There is [iOSSnapshotTestCase](https://github.com/uber/ios-snapshot-test-case/) which works exactly like that for [Texture](https://texturegroup.org).

**Comment by mmcc007** [comment]
Added support for screenshots comparison. 
Give it a try and let me know how it works out:
```
git clone https://github.com/mmcc007/screenshots.git
cd screenshots
git checkout #65_screenshots_comparison
pub global activate --source path .
cd example
screenshots -m recording
screenshots -m comparison
```

**Comment by ildarsharafutdinov** [comment]
@mmcc007 ,

It looks like comparison is incorrect. Test fails:
```
...
Processing screenshots from test...
Moving screenshots to ios/fastlane/screenshots/en-US
Running comparison with recorded screenshots in /tmp/screenshots_record/ios/fastlane/screenshots/en-US ...
Unhandled exception:
Error: ios/fastlane/screenshots/en-US/iPhone X-1.png is not equal to /tmp/screenshots_record/ios/fastlane/screenshots/en-US/iPhone X-1.png
```

However, images are equal(all zeros):
```
magick compare -verbose -metric mae "ios/fastlane/screenshots/en-US/iPhone X-1.png" "/tmp/screenshots_record/ios/fastlane/screenshots/en-US/iPhone X-1.png" difference.png
ios/fastlane/screenshots/en-US/iPhone X-1.png PNG 1125x2436 1125x2436+0+0 8-bit sRGB 60431B 0.490u 0:00.074
/tmp/screenshots_record/ios/fastlane/screenshots/en-US/iPhone X-1.png PNG 1125x2436 1125x2436+0+0 8-bit sRGB 60431B 0.480u 0:00.056
Image: ios/fastlane/screenshots/en-US/iPhone X-1.png
  Channel distortion: MAE
    red: 0 (0)
    green: 0 (0)
    blue: 0 (0)
    alpha: 0 (0)
    all: 0 (0)
ios/fastlane/screenshots/en-US/iPhone X-1.png=>difference.png PNG 1125x2436 1125x2436+0+0 8-bit sRGB 60431B 3.920u 0:00.509
```

Also I would be great if `screenshots` could generate diff image when comparison fails.

**Comment by ildarsharafutdinov** [comment]
Is it possible to support a `tolerance` parameter for `screenshots -m comparison` in order accept image difference to a certain degree?

**Comment by mmcc007** [comment]
Will do some more testing on different devices... to reproduce your results and fix. 

To add more advanced comparison features can switch to using magick (if appropriate). Do you have any suggestions on simple diffs and tolerance implementations using magick or whatever?

**Comment by mmcc007** [comment]
Fixed the comparison using imageMagick and added support for diffs. The screenshot diffs will now show-up in the fastlane dirs with a '-diff' suffix. Try it out and see how it works.

What do you have in mind for tolerances. Do you have an imageMagick example?

**Comment by ildarsharafutdinov** [comment]
I think it's ok to require imagemagick for advanced features of `screenshots`.

By tolerance I mean [ Fuzz Factor](http://www.imagemagick.org/Usage/color_basics/#fuzz) & `fuzz` parameter. 

E.g. here is command and output:

`magick compare -mc AE -fuzz 50% iPhone\ X-0.png iPhone\ X-1.png diff.png`
![diff](https://user-images.githubusercontent.com/1442901/61026260-f1ab5b00-a3c3-11e9-87aa-8eaa649d5470.png)

`5%` leads to the same diff image, but outputs `2082`:

> The special "-metric" setting of 'AE' (short for "Absolute Error" count), will report (to standard error), a count of the actual number of pixels that were masked, at the current fuzz factor.
> ...
> As of IM v6.4.3  the  -metric AE  count is -fuzz effected. so you can discount 'minor' differences from this count.
> ...
> This is the ONLY metric which is 'fuzz' effected.

**Comment by ildarsharafutdinov** [comment]
hmm...

I run `screenshots -m comparison` and it looks like it regenerates images in `ios/fastlane/screenshots/en-US/` with each run.

```
ls -la ios/fastlane/screenshots/en-US/
total 456
drwxr-xr-x  7 ildar  staff    224 Jul 11 10:55 .
drwxr-xr-x  4 ildar  staff    128 Jul 10 12:32 ..
-rw-r--r--@ 1 ildar  staff   6148 Jul 11 10:55 .DS_Store
-rw-r--r--  1 ildar  staff  47895 Jul 11 10:55 iPhone X-0-diff.png
-rw-r--r--@ 1 ildar  wheel  62494 Jul 11 10:55 iPhone X-0.png
-rw-r--r--  1 ildar  staff  46494 Jul 11 10:55 iPhone X-1-diff.png
-rw-r--r--@ 1 ildar  wheel  60431 Jul 11 10:55 iPhone X-1.png
```

```
ls -la ios/fastlane/screenshots/en-US/
total 456
drwxr-xr-x  7 ildar  staff    224 Jul 11 10:57 .
drwxr-xr-x  4 ildar  staff    128 Jul 10 12:32 ..
-rw-r--r--@ 1 ildar  staff   6148 Jul 11 10:55 .DS_Store
-rw-r--r--  1 ildar  staff  47895 Jul 11 10:57 iPhone X-0-diff.png
-rw-r--r--  1 ildar  wheel  62494 Jul 11 10:57 iPhone X-0.png
-rw-r--r--  1 ildar  staff  46494 Jul 11 10:57 iPhone X-1-diff.png
-rw-r--r--  1 ildar  wheel  60431 Jul 11 10:57 iPhone X-1.png
```

`iPhone X-0.png` & `iPhone X-1.png` are reference images and are supposed to be created with `screenshots -m recording` and stay immutable with `screenshots -m comparison`.

**Comment by mmcc007** [comment]
As far as image handling goes, a recording run does the same thing as a comparison/normal run in that it generates the images for each device. The difference being that for a recording run, the images are placed in a separate directory for later comparison, so can be considered immutable until the next recording run for that device.

After a recording run for a device (and any subsequent normal/comparison runs):
```
find /tmp/screenshots_record -name 'iPhone X*' -print0 | xargs -0 ls -la
-rw-r--r--  1 jenkins  wheel  726434 Jul 10 08:57 /tmp/screenshots_record/ios/fastlane/screenshots/en-US/iPhone X-0.png
-rw-r--r--  1 jenkins  wheel  718649 Jul 10 08:57 /tmp/screenshots_record/ios/fastlane/screenshots/en-US/iPhone X-1.png
```
After a normal/comparison run for a device:
```
find example -name 'iPhone X*' -print0 | xargs -0 ls -la
-rw-r--r--  1 jenkins  wheel  726434 Jul 10 18:14 example/ios/fastlane/screenshots/en-US/iPhone X-0.png
-rw-r--r--  1 jenkins  wheel  718649 Jul 10 18:14 example/ios/fastlane/screenshots/en-US/iPhone X-1.png
```
So should be doing the right thing (let me know if otherwise).

BTW: since imageMagick generates a diff image during a comparison run even if there is no difference, I think it makes sense to delete these no-difference diffs during the comparison run and only keep them around if the comparison fails.

**Comment by mmcc007** [comment]
As far as fuzz factor goes, since the images generated during a comparison run are generally exact replicas (per imageMagick) of the images generated during a recording run and are only expected to change if there is a change in the UI of the screenshot, I don't see the utility of adding a fuzz factor. 

Am I missing something?

**Comment by ildarsharafutdinov** [comment]
hi,

> So should be doing the right thing (let me know if otherwise).

I got you. So the steps can be as follows:
- set `recording: <path-in-my-repo>` in `screenshots.yaml`
- run `screenshots -m recording`
- commit dir `<path-in-my-repo>`
- run `screenshots -m comparison` on CI, which outputs `example/ios/fastlane/screenshots/en-US/iPhone X-0.png` along with diff
- in case of tests failure archive `example/ios/fastlane/screenshots/**/*diff.png`

> since imageMagick generates a diff image during a comparison run even if there is no difference, I think it makes sense to delete these no-difference diffs during the comparison run and only keep them around if the comparison fails.

agree

> I don't see the utility of adding a fuzz factor

We experienced difference which is not visible with human eye when Xcode versions differ on developer machine and CI. However, it happened with Texture and now I tend to agree with you. It's probably not worth implementing in advance.

I have one more question though. We currently run our iOS & Android tests in parallel. iOS is run on mac mini node, whereas Android on linux node. This means that we cannot run all the tests since e.g. we cannot run iOS simulator on linux.
Is there a way to tell screenshots "hey, take only ios devices from `screenshots.yaml` and skip the rest"?

**Comment by mmcc007** [comment]
Thanks for the feedback. Since it seems to be working, I will release this.

> Is there a way to tell screenshots "hey, take only ios devices from screenshots.yaml and skip the rest"?

Use 2 config files, like screenshots-ios.yaml and screenshots-android.yaml. If you prefer using only one config file open a new issue and can add an argument like `--device-type [ios/android]` or can add a parameter to config file to detect if running on linux.

**Comment by mmcc007** [comment]
This has been release in v1.1.0
